### PR TITLE
Update az prompted survey mechanism

### PR DESCRIPTION
**Pull request by Nickcandy**
[Update az prompted survey mechanism](https://github.com/Azure/azure-powershell-common/commit/74315e94d457e8eb8da918a66135347c68977fa8)

### CODE REVIEW

File: src/Share/Survey/SurveyHelper.cs [hunk 1]
```diff
@@ -78,213 +97,62 @@ public static SurveyHelper GetInstance()
...
-            foreach(ModuleInfo info in externalScheduleInfo?.Modules)
-            {
-                externalModules[info.Name] = info;
-            }
-
-            HashSet<string> moduleNames = new HashSet<string>(Modules.Keys);
-            moduleNames.UnionWith(new HashSet<string>(externalModules.Keys));
-
-            foreach (string name in moduleNames)
+        private void InitialSurveyHelper(){
```
Comment (isra-fel) [inline src/Share/Survey/SurveyHelper.cs:128]: Our coding style is to put { in a standalone line. Please remember to format the document after making any changes
```suggestion
        private void InitialSurveyHelper()
        {
```

**Review by isra-fel — CHANGES_REQUESTED** [review]
- The installation ID is not strongly tied to survey or telemetry, so I recommend extracting it to a `Share/InstallationHelper.cs`.
- Have you tested the compatibility of new code + old AzureRmSurvey.json?

File: src/Common/MetricHelper.cs [hunk 2]
```diff
@@ -581,6 +582,8 @@ public class AzurePSQoSEvent
     public string InvocationName { get; set; }
     public Dictionary<string, string> CustomProperties { get; private set; }
 
+    public String InstallationId {get; set;}
```
Comment (dingmeng-xue) [inline src/Common/MetricHelper.cs:585]: Is it required?

File: src/Common/AzurePSCmdlet.cs [hunk 3]
```diff
@@ -668,6 +674,8 @@ protected virtual void InitializeQosEvent()
             _qosEvent.PSHostName = PSHostName;
             _qosEvent.ModuleName = this.ModuleName;
             _qosEvent.ModuleVersion = this.ModuleVersion;
+            AzureSession.Instance.ExtendedProperties.TryGetValue("InstallationId", out String InstallationId);
+            _qosEvent.InstallationId = InstallationId;
```
Comment (dingmeng-xue) [inline src/Common/AzurePSCmdlet.cs:678]: Given _qosEvent should have installation at beginning. Why is it set again?

File: src/Share/Survey/SurveyHelper.cs [hunk 4]
```diff
@@ -78,213 +97,62 @@ public static SurveyHelper GetInstance()
             }
             return _instance;
         }
-
-        public bool ShouldPropmtSurvey(string moduleName, Version moduleVersion)
-        {
+        public bool ShouldPropmtAzSurvey(String installationId){
```
Comment (dingmeng-xue) [inline src/Share/Survey/SurveyHelper.cs:100]: Do you need `installationId` to determine survey should be prompt or not? If no, this parameter should be removed.

File: src/Common/AzurePSCmdlet.cs [hunk 5]
```diff
@@ -53,6 +53,10 @@ public abstract class AzurePSCmdlet : PSCmdlet, IDisposable
         protected IList<Regex> _matchers { get;  private set; }  = new List<Regex>();
         private static readonly Regex _defaultMatcher = new Regex("(\\s*\"access_token\"\\s*:\\s*)\"[^\"]+\"");
 
+        private static string ansiCodePrefix = "\u001b[97;48;2;0;120;212m";
```
Comment (dingmeng-xue) [inline src/Common/AzurePSCmdlet.cs:56]: Please add comment

File: src/Common/AzurePSCmdlet.cs [hunk 6]
```diff
@@ -380,7 +384,9 @@ private void WriteBreakingChangeOrPreviewMessage()
         /// </summary>
         protected override void EndProcessing()
         {
-            if (this.MyInvocation?.MyCommand?.Version != null && SurveyHelper.GetInstance().ShouldPropmtSurvey(this.MyInvocation.MyCommand.ModuleName, this.MyInvocation.MyCommand.Version))
+            AzureSession.Instance.ExtendedProperties.TryGetValue("InstallationId", out String InstallationId);
+            if (SurveyHelper.GetInstance().ShouldPromptAzSurvey(InstallationId) && (AzureSession.Instance.TryGetComponent<IConfigManager>(nameof(IConfigManager), out var configManager)
+                    && !configManager.GetConfigValue<bool>(ConfigKeysForCommon.EnableInterceptSurvey).Equals(false)))
```
Comment (dingmeng-xue) [inline src/Common/AzurePSCmdlet.cs:389]: Please correct this line.

File: src/Common/MetricHelper.cs [hunk 7]
```diff
@@ -58,6 +59,8 @@ public class MetricHelper
 
         public static void AppendInternalCalledCmdlet(string cmldetName) { _internalCalledCmdlets += (string.IsNullOrEmpty(_internalCalledCmdlets) ? "" : ",") + cmldetName; }
 
+        public static string InstallationId { get => AzureSession.Instance.ExtendedProperties.TryGetValue("InstallationId", out String InstallationId)?InstallationId:""; }
```
Comment (dingmeng-xue) [inline src/Common/MetricHelper.cs:62]: String.Empty

**Pull request closed without merging**